pep-0544.txt

```diff
@@ -0,0 +1,1006 @@
+they can use the attribute on a particular instance.
+
+Languages such as TypeScript have similar features and
+apparently they are pretty commonly used. The current realistic potential
+use cases for protocols in Python don't require these. In the interest
+of simplicity, we propose to not support optional methods or attributes.
+We can always revisit this later if there is an actual need.
+
+
+Make protocols interoperable with other approaches
+--------------------------------------------------
+
+The protocols as described here are basically a minimal extension to
+the existing concept of ABCs. We argue that this is the way they should
+be understood, instead of as something that *replaces* Zope interfaces,
+for example. Attempting such interoperabilities will significantly
+complicate both the concept and the implementation.
+
+On the other hand, Zope interfaces are conceptually a superset of protocols
+defined here, but using an incompatible syntax to define them,
+because before PEP 526 there was no straightforward way to annotate attributes.
+In the 3.6+ world, ``zope.interface`` might potentially adopt the ``Protocol``
+syntax. In this case, type checkers could be taught to recognize interfaces
+as protocols and make simple structural checks with respect to them.
+
+
+Use assignments to check explicitly that a class implements a protocol
+----------------------------------------------------------------------
+
+In Go language the explicit checks for implementation are performed
+via dummy assignments [golang]_. Such a way is also possible with the
+current proposal. Example::
+
+  class A:
+      def __len__(self) -> float:
+          return ...
+
+  _: Sized = A()  # Error: A.__len__ doesn't conform to 'Sized'
+                  # (Incompatible return type 'float')
+
+This approach moves the check away from
+the class definition and it almost requires a comment as otherwise
+the code probably would not make any sense to an average reader
+-- it looks like dead code. Besides, in the simplest form it requires one
+to construct an instance of ``A``, which could be problematic if this requires
+accessing or allocating some resources such as files or sockets.
+We could work around the latter by using a cast, for example, but then
+the code would be ugly. Therefore we discourage the use of this pattern.
+
+
+Support ``isinstance()`` checks by default
+------------------------------------------
+
+The problem with this is instance checks could be unreliable, except for
+situations where there is a common signature convention such as ``Iterable``.
+For example::
+
+  class P(Protocol):
+      def common_method_name(self, x: int) -> int: ...
+
+  class X:
+      <a bunch of methods>
+      def common_method_name(self) -> None: ... # Note different signature
+
+  def do_stuff(o: Union[P, X]) -> int:
+      if isinstance(o, P):
+          return o.common_method_name(1)  # oops, what if it's an X instance?
```

Comment:
> Or are you saying that in this case if ``P`` was decorated with ``@runtime`` the isinstance check would still be broken?

I think yes, for example now ``isinstance(obj, Iterable)`` returns ``True`` for a class with ``__iter__ = 42``. And we can't make ``@runtime`` much better because not all information is available statically. As Jukka said, it is not a big problem since static type checker will warn about this, but big enough to make support for ``isinstance()`` with protocols opt-in.